lib/bson/array.rb

```diff
@@ -93,13 +97,17 @@ module ClassMethods
       #
       # @since 2.0.0
       def from_bson(buffer)
-        array = new
-        buffer.get_int32 # throw away the length
-        while (type = buffer.get_byte) != NULL_BYTE
-          buffer.get_cstring
-          array << BSON::Registry.get(type).from_bson(buffer)
+        if buffer.respond_to?(:get_array)
+          buffer.get_array
+        else
+          array = new
+          buffer.get_int32 # throw away the length
```

Comment:
I realize the original throws away size, but as I mentioned, I'd prefer to see validation.